### ISSUE TITLE
Min Needs Calculator: py error and error handling

### ISSUE THREAD

**Issue by Charlotte-Morgan** [inasafe/inasafe#3498]
# problem
the min needs calculator returns an error when supplied with data containing 0 and NULL values

# error message 
```
An error has occurred while executing Python code: 

WriteLayerError: Creation of output file C:/AIFDR/AIFDR II - SDA/00_20161207_AcehEarthquake/1_Data/Support data/IDP/20161216_0900_IDP_MinNeeds_perka7.shp failed 
Traceback (most recent call last):
  File "C:/Users/AIFDR_ADMIN/.qgis2/python/plugins\inasafe\safe\gui\tools\minimum_needs\needs_calculator_dialog.py", line 209, in accept
    output_layer.write_to_file(new_file)
  File "C:/Users/AIFDR_ADMIN/.qgis2/python/plugins\inasafe\safe\storage\vector.py", line 679, in write_to_file
    raise WriteLayerError(msg)
WriteLayerError: Creation of output file C:/AIFDR/AIFDR II - SDA/00_20161207_AcehEarthquake/1_Data/Support data/IDP/20161216_0900_IDP_MinNeeds_perka7.shp failed


Python version: 2.7.5 (default, May 15 2013, 22:44:16) [MSC v.1500 64 bit (AMD64)] 
QGIS version: 2.14.3-Essen Essen, cf2ebb8 

Python Path:
C:/Users/AIFDR_ADMIN/.qgis2/python/plugins\processing
C:/PROGRA~1/QGISES~1/apps/qgis/./python
C:/Users/AIFDR_ADMIN/.qgis2/python
C:/Users/AIFDR_ADMIN/.qgis2/python/plugins
C:/PROGRA~1/QGISES~1/apps/qgis/./python/plugins
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\matplotlib-1.3.1-py2.7-win-amd64.egg
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\nose-1.3.3-py2.7.egg
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\tornado-4.0.1-py2.7-win-amd64.egg
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\backports.ssl_match_hostname-3.4.0.2-py2.7.egg
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\certifi-14.05.14-py2.7.egg
c:\osgeo4~1\apps\python27\lib\site-packages\python_dateutil-2.1-py2.7.egg
c:\osgeo4~1\apps\python27\lib\site-packages\six-1.3.0-py2.7.egg
C:\Program Files (x86)\ArcGIS\Desktop10.2\bin
C:\Program Files (x86)\ArcGIS\Desktop10.2\arcpy
C:\Program Files (x86)\ArcGIS\Desktop10.2\ArcToolbox
C:\PROGRA~1\QGISES~1\bin\python27.zip
C:\PROGRA~1\QGISES~1\apps\Python27\DLLs
C:\PROGRA~1\QGISES~1\apps\Python27\lib
C:\PROGRA~1\QGISES~1\apps\Python27\lib\plat-win
C:\PROGRA~1\QGISES~1\apps\Python27\lib\lib-tk
C:\PROGRA~1\QGISES~1\bin
C:\PROGRA~1\QGISES~1\apps\Python27
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\PIL
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\jinja2-2.7.2-py2.7.egg
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\markupsafe-0.23-py2.7-win-amd64.egg
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\pytz-2012j-py2.7.egg
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\win32
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\win32\lib
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\Pythonwin
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\Shapely-1.2.18-py2.7-win-amd64.egg
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\wx-2.8-msw-unicode
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\xlrd-0.9.2-py2.7.egg
C:\PROGRA~1\QGISES~1\apps\Python27\lib\site-packages\xlwt-0.7.5-py2.7.egg
C:/Users/AIFDR_ADMIN/.qgis2//python
C:/Users/AIFDR_ADMIN/.qgis2/python/plugins\inasafe
C:\Users\AIFDR_ADMIN\.qgis2\python\plugins\inasafe\safe_extras\parameters
C:\Users\AIFDR_ADMIN\.qgis2\python\plugins
C:\Users\AIFDR_ADMIN\.qgis2\python\plugins\inasafe\safe_extras
C:\PROGRA~1\QGISES~1\apps\qgis\python\plugins\fTools\tools
C:/AIFDR/AIFDR II - SDA/00_20161207_AcehEarthquake/2_Analysis/Population
```
# environment
QGIS 2.14.3
InaSAFE 3.5.3
Windows

# sample data
[20161216_0900_IDP_MinNeeds.zip](https://github.com/inasafe/inasafe/files/656564/20161216_0900_IDP_MinNeeds.zip)

# proposed solution 1
the min need calculator can handle real world data
if a 0 population is input it should return a 0 output
if a NULL value is input it should return a NULL output

# proposed solution 2
The min needs calculator returns a happy message to tell users what is wrong with input data

**Comment by timlinux** [comment]
@ismailsunni can you please take a look?

**Comment by ismailsunni** [comment]
Hi @Charlotte-Morgan 
I tried the data but couldn't replicate the error. I have also tried with Null and 0 data, but it still give me fine result. The Null is calculated as 0, but no error message.

I will make a fix to return Null for Null value and put it in 3.5.5 release.

Looking at your path of file, there might be a permission problem. Can you try it again after copying the file to another place?

**Comment by Charlotte-Morgan** [comment]
Hi @ismailsunni  - my apologies, I thought I shared the data that caused the problem in the original ticket.

Please try this one.  The IDP count is in the field IDP_16Dec
[IDP_SubDistrictReports.zip](https://github.com/inasafe/inasafe/files/659845/IDP_SubDistrictReports.zip)

**Comment by ismailsunni** [comment]
Hi @Charlotte-Morgan 

I have tried that one, it gives me no error but all null set to zero.

**Comment by Charlotte-Morgan** [comment]
@felix-yew  - can you please take a look at this and see if you can replicate the error I reported - ie do you get a python error?

**Comment by felix-yew** [comment]
Hi @Charlotte-Morgan and mas @ismailsunni 

I try the IDP data from Charlotte (IDP_SubDistrictReports.zip) and run it with InaSAFE 3.5.3 and QGIS 2.14.7 get some error
```
An error has occurred while executing Python code: 
ReadLayerError: Geometry was None in filename D:/Felix/Test/Minimum Needs Calculator/IDP_SubDistrictReports/IDP_SubDistrictReports.shp 
Traceback (most recent call last):
  File "C:/Users/yanuaf/.qgis2/python/plugins\inasafe\safe\gui\tools\minimum_needs\needs_calculator_dialog.py", line 200, in accept
    input_layer = safe_read_layer(file_name)
  File "C:/Users/yanuaf/.qgis2/python/plugins\inasafe\safe\storage\core.py", line 30, in read_layer
    return Vector(filename)
  File "C:/Users/yanuaf/.qgis2/python/plugins\inasafe\safe\storage\vector.py", line 165, in __init__
    self.read_from_file(data)
  File "C:/Users/yanuaf/.qgis2/python/plugins\inasafe\safe\storage\vector.py", line 488, in read_from_file
    raise ReadLayerError(msg)
ReadLayerError: Geometry was None in filename D:/Felix/Test/Minimum Needs Calculator/IDP_SubDistrictReports/IDP_SubDistrictReports.shp 

Python version: 2.7.5 (default, May 15 2013, 22:44:16) [MSC v.1500 64 bit (AMD64)] 
QGIS version: 2.14.7-Essen Essen, ed8807c 
Python Path:
•	C:/PROGRA~1/QGIS2~1.14/apps/qgis-ltr/./python/plugins\processing
•	C:/PROGRA~1/QGIS2~1.14/apps/qgis-ltr/./python
•	C:/Users/yanuaf/.qgis2/python
•	C:/Users/yanuaf/.qgis2/python/plugins
•	C:/PROGRA~1/QGIS2~1.14/apps/qgis-ltr/./python/plugins
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\matplotlib-1.3.1-py2.7-win-amd64.egg
•	C:\Users\yanuaf\.qgis2\python\plugins\geoserverexplorer\ext-libs\httplib2-0.8-py2.7.egg
•	C:\Users\yanuaf\.qgis2\python\plugins\geoserverexplorer\ext-libs\python_dateutil-2.2-py2.7.egg
•	C:\Users\yanuaf\.qgis2\python\plugins\geoserverexplorer\ext-libs\six-1.10.0-py2.7.egg
•	C:\Users\yanuaf\.qgis2\python\plugins\geoserverexplorer\ext-libs\requests-2.5.0-py2.7.egg
•	C:\Users\yanuaf\.qgis2\python\plugins\geoserverexplorer\ext-libs\gsconfig-1.0.6-py2.7.egg
•	C:\Users\yanuaf\.qgis2\python\plugins\geoserverexplorer\ext-libs\gisdata-0.5.4-py2.7.egg
•	C:\Users\yanuaf\.qgis2\python\plugins\geoserverexplorer\ext-libs\gsimporter-1.0.0-py2.7.egg
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\nose-1.3.3-py2.7.egg
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\tornado-4.0.1-py2.7-win-amd64.egg
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\backports.ssl_match_hostname-3.4.0.2-py2.7.egg
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\certifi-14.05.14-py2.7.egg
•	C:\PROGRA~1\QGIS2~1.14\bin\python27.zip
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\DLLs
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\plat-win
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\lib-tk
•	C:\PROGRA~1\QGIS2~1.14\bin
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\PIL
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\jinja2-2.7.2-py2.7.egg
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\markupsafe-0.23-py2.7-win-amd64.egg
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\pytz-2012j-py2.7.egg
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\win32
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\win32\lib
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\Pythonwin
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\Shapely-1.2.18-py2.7-win-amd64.egg
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\wx-2.8-msw-unicode
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\xlrd-0.9.2-py2.7.egg
•	C:\PROGRA~1\QGIS2~1.14\apps\Python27\lib\site-packages\xlwt-0.7.5-py2.7.egg
•	C:/Users/yanuaf/.qgis2//python
•	C:\Users\yanuaf\.qgis2\python\plugins\geoserverexplorer\ext-libs
•	C:/Users/yanuaf/.qgis2/python/plugins\inasafe
•	C:\Users\yanuaf\.qgis2\python\plugins\inasafe\safe_extras\parameters
•	C:\Users\yanuaf\.qgis2\python\plugins
•	C:\Users\yanuaf\.qgis2\python\plugins\inasafe\safe_extras
•	C:\Users\yanuaf\.qgis2\python\plugins\mmqgis/forms
•	C:\PROGRA~1\QGIS2~1.14\apps\qgis-ltr\python\plugins\fTools\tools
```

But, if I run this data with InaSAFE 3.5.4 and QGIS 2.14.7, it has a result like this:
![image](https://cloud.githubusercontent.com/assets/13621886/21301986/ffeca972-c5e4-11e6-8b53-e71b62ad4abb.png)

**Comment by Charlotte-Morgan** [comment]
@felix-yew  - you are telling me its fixed in 3.5.4?
That is good :)

**Comment by Charlotte-Morgan** [comment]
@ismailsunni  - i can confirm that in 3.5.4 the python error is fixed. I would still prefer that if input is NULL then output is NULL and if input is 0 output is 0

**Comment by ismailsunni** [comment]
@Charlotte-Morgan yes, agree. I am investigating to make NULL return NULL now. It's harder than I imagine due to the custom vector layer from Ole :)

**Comment by Charlotte-Morgan** [comment]
hi @timlinux - this was related to the Min Needs tool

**Comment by timlinux** [comment]
@myarjunar For you to look at when you are looking at minimum needs....

**Comment by myarjunar** [comment]
closed in #3645
